File: DIGRAPH/bc-tennis.pl
#!/bin/perl

# script to create state diagram for tennis game

require "/usr/local/lib/bclib.pl";

# scores:
# 0 = 0
# 1 = 15
# 2 = 30
# 3 = duece w disadvantage
# 4 = 40 (or deuce)
# 5 = deuce w advantage
# 6 = 60 (victory)

# if x has x points, y has y points, and n scores (n=0 -> x, n=1 ->y),
# return next score in x,y format

sub nextscore {
  my(@l) = @_;
  my($n) = $l[2];
  debug("nextscore(@l)");

  # 15 points or less just increases score, and duece with advantage wins
  if ($l[$n]<=2 || $l[$n]==5) {
    # jump from 30 to 40
    if ($l[$n]==2) {$l[$n]+=2;} else {$l[$n]++;}
    return @l[0..1];
  }

  # if you have disadvantage and score, go to deuce
  if ($l[$n] == 3) {return (4,4);}

  # you have 40, opponent has 30 or less -> you win
  if ($l[$n]==4 && $l[1-$n]<=2) {
    $l[$n] = 6;
    return @l[0..1];
  }

  # deuce
  if ($l[$n]==4 && $l[1-$n]==4) {
    $l[$n] = 5;
    $l[1-$n] = 3;
    return @l[0..1];
  }

  return "NIL";
}

# fill in the state diagram starting with score of $x,$y (recursion)

sub statescore {
  my($x,$y) = @_;
  debug("statescore($x,$y)");

  # if either side has won, nothing more to do
  if ($x==6 || $y==6) {return;}

  # fill in hash for 0 and 1 values and recurse unless I already have them
  unless ($statehash{$x}{$y}{0}) {
    $statehash{$x}{$y}{0} = nextscore($x,$y,0);
    statescore(nextscore($x,$y,0));
  }

  unless ($statehash{$x}{$y}{1}) {
    $statehash{$x}{$y}{1} = nextscore($x,$y,1);
    statescore(nextscore($x,$y,1));
  }
}


statescore(0,0);

die "TESTING";

@scores = (0,0);


debug(nextscore(@scores,0));
debug(nextscore(@scores,1));

die "TESTING";

while (@scores) {
  push(@scores, nextscore(@scores,0));
  push(@scores, nextscore(@scores,1));
  debug("SCORES",@scores);
}

File: DIGRAPH/bc-tennis.rb
#!/usr/local/bin/ruby

# script to create state diagram for tennis game

require '/home/barrycarter/BCGIT/bclib.rb'

# scores:
# 0 = 0
# 1 = 15
# 2 = 30
# 3 = deuce w disadvantage
# 4 = 40 (or deuce)
# 5 = deuce w advantage
# 6 = 60 (victory)

# GameState is a list; entry n = player n score (n=0|1)

class GameState
 def initialize(l) @l=l end

 # TODO: genericize this
 def l() @l end

 def inspect() return @l end

 # the next state if nth player scores
 def nextstate(n)
   # store current state in array
   t = @l.dup;

   # one player has won? nothing more to do, but return empty game
   if (@l[0]==6 || @l[1]==6) then return self end

   # player with 0 or 15 or deuce advantage scores
   if @l[n]<=1 || @l[n]==5 then
     t[n]+=1;
     return GameState.new(t) 
   end

   # player with 40 scores, other player has 30 or less
   if @l[n]==4 && @l[1-n]<=2 then
     t[n] = 6;
     return GameState.new(t)
   end

   # player with 30 scores
   if @l[n]==2 then
     t[n] = 4;
     return GameState.new(t)
   end

   # player with deuce disadvantage scores (yields deuce)
   if @l[n]==3 then return GameState.new([4,4]) end

   # during deuce, one player scores
   if @l[n]==4 && @l[1-n]==4 then
     t[n] = 5;
     t[1-n] = 3;
     return GameState.new(t)
   end

   self.td("NO NEXT STATE?")

 end
end

# start with an array containing new game
$DEBUG=1;
games = [GameState.new([0,0])]
hash = Hash.new
seen = Hash.new

while i=games.shift do
  if i==0 then next end
  # have we seen this game (ie, this pair of scores)
  if seen[i.l] then next else seen[i.l]=1 end
  td("NEXT STATES: #{i.l} -> #{i.nextstate(0).l}, #{i.nextstate(1).l}")
  hash[i] = Hash.new
  hash[i][0] = i.nextstate(0)
  hash[i][1] = i.nextstate(1)
  games.push(i.nextstate(0))
  games.push(i.nextstate(1))
#  games.inspect.td("GAMES")
end

# print hash.keys.each{|i| print "#{i}, #{hash[i][0]}, #{hash[i][1]}\n"}

# print GameState.new([4,2]).nextstate(1).inspect
# hash = Hash.new(Hash.new)








